### PR TITLE
`sage.matroids`: Modularization fixes

### PR DESCRIPTION
**Pull request by mkoeppe**
### :books: Description

Fixes for imports and adding `# optional`.

Also some `# optional` updates in other modules.

- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by tobiasdiez** [comment]
Would it be possible to filter out all these optional tags from the docs? Its a bit annoying and they don't really add anything.

**Comment by mkoeppe** [comment]
> Would it be possible to filter out all these optional tags from the docs?

That would be very useful for the PDF documentation because in that format, the comments wrap around.

For HTML, I want to keep them.

> Its a bit annoying 

 I have aligned these annotations so that only a little bit (`# option`....) is visible in order to minimize annoyance.

> and they don't really add anything.

They do add important information for users of the modularized distributions: They can see in the documentation what examples can be run when this or that is installed but other things aren't, so that they can avoid trial and error.

Moreover, annotations that refer to libraries, such as `# optional - sage.libs.singular`, also give attribution to where the functionality of this example is implemented.

Here in the case of `sage.matroids`, the annotations also give the reader a rough idea how examples are constructed (from graphs? from linear algebra over finite fields? from schemes?), and which parts of the functionality are graph-theoretic is nature.

**Comment by tobiasdiez** [comment]
> > Its a bit annoying
> 
> I have aligned these annotations so that only a little bit (`# option`....) is visible in order to minimize annoyance.
> 

This is not really working for me.
image


> > and they don't really add anything.
> 
> They do add important information for users of the modularized distributions: They can see in the documentation what examples can be run when this or that is installed but other things aren't, so that they can avoid trial and error.

In my opinion this "# optional" tag is not self-explanatory enough to provide the information you describe. I agree that users might profit from this, but currently its too opaque. I propose we strip them for now, and if you think the information is worth it parse the optional tags and add a tiny notice "This example requires xyz and abc" at the end, with hyperlinks to xyz and abc. 

(sorry for this somewhat off-topic discussion, but I think the representation of these optional tags should be discussed before adding even more to the docs.)

**Comment by mkoeppe** [comment]
> In my opinion this "# optional" tag is not self-explanatory enough to provide the information you describe. I agree that users might profit from this, but currently its too opaque.

Sure, documentation will be needed. It will be added in the documentation of `sage.features` and in the READMEs of the distributions. (See https://github.com/sagemath/sage/pull/35095/files for the new distributions.)

**Comment by mkoeppe** [comment]
> I propose we strip them for now, and if you think the information is worth it parse the optional tags and add a tiny notice

I don't plan to work on this

**Comment by tobiasdiez** [comment]
Based on https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-trim_doctest_flags and https://github.com/sphinx-doc/sphinx/blob/d3c91f951255c6729a53e38c895ddc0af036b5b9/sphinx/transforms/post_transforms/code.py#L82-L125 this shouldn't be that hard. Maybe
```py
from sphinx.ext import doctest as sphinx_doctest
sphinx_doctest.doctestopt_re = re.compile(r'#\s*(doctest|optional):.+$', re.MULTILINE)
trim_doctest_flags = True
```
or overwriting `TrimDoctestFlagsTransform`

**Comment by mkoeppe** [comment]
Thanks, I've added this info to:
- #35726

That we are adding these annotations is already documented in https://doc.sagemath.org/html/en/developer/packaging_sage_library.html#preparing-doctests

There's nothing that "needs work" here.

**Comment by mkoeppe** [comment]
> > In my opinion this "# optional" tag is not self-explanatory enough to provide the information you describe. I agree that users might profit from this, but currently its too opaque.
> 
> Sure, documentation will be needed. It will be added in the documentation of `sage.features` 

This is happening in #35749

**Comment by mkoeppe** [comment]
The test failure in ell_rational_field is unrelated.

File: src/sage/dynamics/arithmetic_dynamics/projective_ds.py [hunk 1]
```diff
@@ -1866,7 +1869,7 @@ def conjugate(self, M, adjugate=False, normalize=False):
 
         - ``M`` -- a square invertible matrix
 
-        - ``adjugate`` -- (default: ``False``) boolean, also classically called adjoint, takes a square matrix ``M`` and finds the transpose of its cofactor matrix. Used for conjugation in place of inverse when specified ``'True'``. Functionality is the same in projective space.
+        - ``adjugate`` -- (default: ``False``) boolean, also classically called adjoint, takes a square matrix ``M`` and finds the transpose of its cofactor matrix. Used for conjugation in place of inverse when specified ``True``. Functionality is the same in projective space.
```
Comment (dcoudert) [inline src/sage/dynamics/arithmetic_dynamics/projective_ds.py]: alignment in 80 columns mode would be nice ;)

File: src/sage/matroids/linear_matroid.pyx [hunk 2]
```diff
@@ -5038,7 +5042,7 @@ cdef class QuaternaryMatroid(LinearMatroid):
         EXAMPLES::
 
             sage: from sage.matroids.advanced import *
-            sage: QuaternaryMatroid(matrix=Matrix(GF(4, 'x'),                 # indirect doctest
+            sage: QuaternaryMatroid(matrix=Matrix(GF(4, 'x'),       # indirect doctest  # optional - sage.rings.finite_rings
```
Comment (dcoudert) [inline src/sage/matroids/linear_matroid.pyx]: could be `# indirect doctest, optional - sage.rings.finite_rings`

File: src/sage/matroids/minor_matroid.py [hunk 3]
```diff
@@ -132,9 +132,9 @@ def __init__(self, matroid, contractions=None, deletions=None):
         EXAMPLES::
 
             sage: from sage.matroids.advanced import *
-            sage: M = MinorMatroid(matroids.named_matroids.Fano(),            # indirect doctest
+            sage: M = MinorMatroid(matroids.named_matroids.Fano(),  # indirect doctest  # optional - sage.rings.finite_rings
```
Comment (dcoudert) [inline src/sage/matroids/minor_matroid.py]: Could be `# indirect doctest, optional - sage.rings.finite_rings`

**Review by dcoudert — CHANGES_REQUESTED** [review]
a few comments.

**Review by dcoudert — APPROVED** [review]
LGTM.

**Comment by mkoeppe** [comment]
Thank you! May I set to "positive review"?

**Comment by dcoudert** [comment]
Done.

**Pull request merged**